viite-UI/src/view/ProjectForm.js

```diff
@@ -650,13 +653,33 @@
         } else {
           cancelChanges();
         }
-        eventbus.trigger("roadAddressProject:startAllInteractions");
+        eventbus.trigger('roadAddressProject:startAllInteractions');
       });
 
       rootElement.on('click', '#saveAndCancelDialogue', function (eventData) {
-        var defaultPopupMessage = 'Haluatko tallentaa tekemäsi muutokset?';
-        displayCloseConfirmMessage(defaultPopupMessage, true);
+        if (currentProject.isDirty) {
+            new GenericConfirmPopup('Haluatko tallentaa tekemäsi muutokset?', {
+                successCallback: function () {
+                    if (!disabledInput) {
+                        createOrSaveProject();
+                        eventbus.once('roadAddress:projectSaved', function () {
+                            _.defer(function () {
+                                closeProjectMode(true);
+                            });
+                        });
```

Comment:
If it's the `createOrSaveProject `that triggers the `roadAddress:projectSaved` it should be registed before we call method.